samza-kv-rocksdb/src/main/scala/org/apache/samza/storage/kv/RocksDbKeyValueStore.scala

```diff
@@ -203,6 +204,18 @@ class RocksDbKeyValueStore(
     new RocksDbIterator(iter)
   }
 
+  def iterate(from: Array[Byte], to: Array[Byte]): KeyValueIterable[Array[Byte], Array[Byte]] = {
+    //snapshot the iterator
+    val iter : RocksDbRangeIterator = range(from, to).asInstanceOf[RocksDbRangeIterator]
+    new KeyValueIterable[Array[Byte], Array[Byte]] {
+      def iterator(): KeyValueIterator[Array[Byte], Array[Byte]] = {
+        // reset to the beginning
+        iter.seek(from)
```

Comment:
I do have a concern with re-purposing the Iterator returned by `range()` and using it to mint an `Iterable` here. For example, Would thread-2 calling `getIterator` end-up resetting the cursor for thread-1? Please let me know.

KeyValueIterable iterable = store.iterate();
**thread-1:** 
 Iterator i1 = iterable.getIterator();
 iterate using i1; 
**thread-2:**
 Iterator i2 = iterable.getIterator();
 iterate using i2;

Alternately, RocksDb has a top-level notion of a 'snapshot()'. Does the JNI binding for RocksDb expose it? IIUC, it seems to be achieving exactly what we want.

Reply:
NO multithreading in a task in this use case.

Reply:
I didn't know about snapshot. Right now when I created the iterator, it does the snapshot at that point and it fits the purpose (single thread rocksDb access). To use snapshot, I think I need to do some benchmarking to understand the performance aspect of it. I am not comfortable using it directly.

Reply:
Looks like we should be able to make this logic work for normal use cases, basically the logic should be:

if (!iter.isClosed() || iter.next() == false) {
    // if the cached iterator already goes to the end, we can reset it for reuse
   iter.seek(from)
   iter
} else {
   // in case the cached iter is not done or already closed, create a new one.
   range(from, to)
}

Looks like this solution should be okey. As a side note, Flink does similar things for their entries() iterable: https://github.com/apache/flink/blob/master/flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java, line 161.